File: android/app/src/main/java/com/openburnbar/data/square/ThreadInboxItem.kt
package com.openburnbar.data.square

// MARK: - Thread Inbox Item (Android parity)

data class ThreadInboxItem(
    val id: String,
    val agentURI: String,
    val title: String,
    val preview: String,
    val lastActivityAtEpoch: Long,
    val unreadCount: Int = 0,
    val needsAttention: Boolean = false,
    val source: Source,
    val liveMissionID: String? = null,
    val searchText: String = listOf(title, preview, agentURI).joinToString(" "),
    val customTitle: String? = null,
    val labelColorHex: String? = null,
    val isPinned: Boolean = false,
    val priorityOrder: Int? = null
) {
    enum class Source(val token: String) {
        HERMES("hermes"),
        PI("pi"),
        CLI_MIRROR("cli_mirror"),
        MISSION_GROUP("mission_group"),
        SUBSCRIPTION_POST("subscription_post");

        companion object {
            fun fromToken(value: String?): Source =
                values().firstOrNull { it.token == value } ?: HERMES
        }
    }
}

fun List<ThreadInboxItem>.sortedForInbox(): List<ThreadInboxItem> =
    sortedWith(Comparator { a, b ->
        // 1. Pinned first
        if (a.isPinned != b.isPinned) {
            return@Comparator if (a.isPinned) -1 else 1
        }
        // 2. Needs attention second
        if (a.needsAttention != b.needsAttention) {
            return@Comparator if (a.needsAttention) -1 else 1
        }
        // 3. priorityOrder ascending (where smaller assigned values rank higher than unassigned null/0)
        val pA = a.priorityOrder?.takeIf { it > 0 }
        val pB = b.priorityOrder?.takeIf { it > 0 }
        if (pA != pB) {
            if (pA != null && pB != null) {
                return@Comparator pA.compareTo(pB)
            } else if (pA != null) {
                return@Comparator -1
            } else if (pB != null) {
                return@Comparator 1
            }
        }
        // 4. lastActivityAtEpoch recency descending
        return@Comparator b.lastActivityAtEpoch.compareTo(a.lastActivityAtEpoch)
    })

fun List<ThreadInboxItem>.splitForInbox(): Pair<List<ThreadInboxItem>, List<ThreadInboxItem>> {
    val service = mutableListOf<ThreadInboxItem>()
    val subscription = mutableListOf<ThreadInboxItem>()
    for (item in this) {
        when (item.source) {
            ThreadInboxItem.Source.SUBSCRIPTION_POST -> subscription.add(item)
            else -> service.add(item)
        }
    }
    return service.sortedForInbox() to subscription.sortedForInbox()
}

File: OpenBurnBarCore/Sources/OpenBurnBarCore/SharedModels/ThreadInboxItem.swift
import Foundation

// MARK: - Thread Inbox Item (Hermes Square §6.2)
//
// Unified view-model for an entry in the Living Inbox. The
// `ThreadInboxStore` aggregates native mobile chat history, Mac-mirrored CLI
// Firestore sessions, missions, and subscription posts into a single list of
// these items.
//
// Plain value type — no protocol identity, no Firebase, no service refs.
// The store builds these as projections of the per-runtime authoritative
// stores.

public struct ThreadInboxItem: Sendable, Hashable, Identifiable {
    /// Stable identity, namespaced by source kind to avoid collisions.
    /// Example: `hermes:abc123`, `pi:xyz789`, `cli:claude-sess-001`,
    /// `subscription:research-scout:weekly-recap`.
    public let id: String

    /// The agent that owns the thread (e.g., `agent://burnbar/claude`).
    public let agentURI: String

    /// Friendly title shown in the list. Usually the first user message
    /// trimmed, or "(untitled)" if empty.
    public let title: String

    /// One-line preview of the most recent message or status.
    public let preview: String

    /// Non-visual text indexed by Hermes Square search. CLI mirrors use
    /// this to include transcript turns without stuffing long text into
    /// the visible preview row.
    public let searchText: String

    /// Date the thread was last touched (sort key).
    public let lastActivityAt: Date

    /// Number of unread items on this thread.
    public let unreadCount: Int

    /// True if this thread has something the user needs to do — typically
    /// an awaiting-approval or failed-mission state.
    public let needsAttention: Bool

    /// What kind of source produced this item — drives the tap behaviour
    /// (open the right native list view).
    public let source: Source

    /// Optional missionID when the item is a live mission. Lets the list
    /// row show a mission tile inline.
    public let liveMissionID: String?

    public enum Source: String, Codable, Sendable, Hashable {
        case hermes
        case pi
        case cliMirror = "cli_mirror"
        case missionGroup = "mission_group"
        case subscriptionPost = "subscription_post"
    }

    public init(
        id: String,
        agentURI: String,
        title: String,
        preview: String,
        lastActivityAt: Date,
        unreadCount: Int = 0,
        needsAttention: Bool = false,
        source: Source,
        liveMissionID: String? = nil,
        searchText: String? = nil
    ) {
        self.id = id
        self.agentURI = agentURI
        self.title = title
        self.preview = preview
        self.searchText = searchText ?? [title, preview, agentURI].joined(separator: " ")
        self.lastActivityAt = lastActivityAt
        self.unreadCount = unreadCount
        self.needsAttention = needsAttention
        self.source = source
        self.liveMissionID = liveMissionID
    }
}

// MARK: - Grouping

extension Array where Element == ThreadInboxItem {
    /// Sort by attention first (needs-attention rows on top), then by
    /// recency. Matches the plan §3 diagram's "needs-attention" sort.
    public func sortedForInbox() -> [ThreadInboxItem] {
        sorted { a, b in
            if a.needsAttention != b.needsAttention {
                return a.needsAttention && !b.needsAttention
            }
            return a.lastActivityAt > b.lastActivityAt
        }
    }

    /// Split into (service-tier, subscription-tier) by source kind. The
    /// inbox renders service-tier above and subscription-tier in the
    /// collapsed folder.
    public func splitForInbox() -> (service: [ThreadInboxItem], subscription: [ThreadInboxItem]) {
        var service: [ThreadInboxItem] = []
        var subscription: [ThreadInboxItem] = []
        for item in self {
            switch item.source {
            case .subscriptionPost:
                subscription.append(item)
            default:
                service.append(item)
            }
        }
        return (service.sortedForInbox(), subscription.sortedForInbox())
    }
}
